src/sas/qtgui/Calculators/Shape2SAS/Constraints.py

```diff
@@ -55,75 +53,297 @@ def __init__(self, parent=None):
+                not isinstance(constraint.value, ast.Name)):
+                return constraint
+
+            lhs = constraint.targets[0].id
+            rhs = constraint.value.id
+
+            # check if lhs is a COM parameter (with or without 'd' prefix)
+            lhs_is_delta = lhs.startswith('d')
+            lhs_base = lhs[1:] if lhs_is_delta else lhs
+
+            if not (lhs_base.startswith("COM") and lhs_base[3:].isdigit()):
+                return constraint
+
+            # check rhs
+            rhs_is_delta = rhs.startswith('d')
+            rhs_base = rhs[1:] if rhs_is_delta else rhs
+
+            new_targets, new_values = [], []
+            if rhs_base.startswith("COM") and rhs_base[3:].isdigit():
+                # rhs is also a COM parameter: COM2 = COM1 -> COMX2, COMY2, COMZ2 =COMX1, COMY1, COMZ1
+                lhs_shape_num = lhs_base[3:]
+                rhs_shape_num = rhs_base[3:]
+
+                for axis in ['X', 'Y', 'Z']:
+                    lhs_full = f"{'d' if lhs_is_delta else ''}COM{axis}{lhs_shape_num}"
+                    rhs_full = f"{'d' if rhs_is_delta else ''}COM{axis}{rhs_shape_num}"
+                    new_targets.append(ast.Name(id=lhs_full, ctx=ast.Store()))
+                    new_values.append(ast.Name(id=rhs_full, ctx=ast.Load()))
 
-        #Get and check translation
-        translation, checkedPars = self.getTranslation(constraintsStr, importStatement, modelPars, modelVals, checkedPars)
+            else:
+                # rhs is a regular parameter: COM2 = X -> COMX2, COMY2, COMZ2 = X, X, X
+                lhs_shape_num = lhs_base[3:]
+                rhs_full = f"{'d' if rhs_is_delta else ''}{rhs_base}"
+                for axis in ['X', 'Y', 'Z']:
+                    lhs_full = f"{'d' if lhs_is_delta else ''}COM{axis}{lhs_shape_num}"
+                    new_targets.append(ast.Name(id=lhs_full, ctx=ast.Store()))
+                    new_values.append(ast.Name(id=rhs_full, ctx=ast.Load()))
+
+            constraint.targets = [ast.Tuple(elts=new_targets, ctx=ast.Store())]
+            constraint.value = ast.Tuple(elts=new_values, ctx=ast.Load())
+            return constraint
+
+        def parse_ast(tree: ast.AST):
+            params = None
+            imports = []
+            constraints = []
 
-        return importStatement, parameters, translation, checkedPars
+            for node in ast.walk(tree):
+                match node:
+                    case ast.ImportFrom() | ast.Import():
+                        imports.append(node)
+
+                    case ast.Assign():
+                        if len(node.targets) != 1 or isinstance(node.targets[0], ast.Tuple) or isinstance(node.value, ast.Tuple):
+                            self.log_embedded_error(f"Tuple assignment is not supported (line {node.lineno}).")
+                            raise ValueError(f"Tuple assignment is not supported (line {node.lineno}).")
```

Comment:
This and a number of other errors are both logged and thrown. These could all be simplified using a functools wrapper - https://stackoverflow.com/questions/11420464/catch-exceptions-inside-a-class.